### PR TITLE
Collection of fixes to test against Istanbul fork

### PR DESCRIPTION
**Pull request by cburgdorf**
### What was wrong?

As noticed in #1748, we do currently not run JSON RPC tests against istanbul fixtures.

### How was it fixed?

1. Ensure `eth_getTransactionReceipt` returns checksum address for `to` field
2. Add incorrect upstream tests (same as Py-EVM)
3. Do some extra normalizations on the fixtures (e.g. casing is sometimes off for `code` or `input` unless there's some checksum scheme I'm unaware of)
4. Activate istanbul CI job

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [x] Clean up commit history

[//]: # (For important changes that should go into the release notes please add a newsfragment file as explained here: https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

[//]: # (See: https://trinity-client.readthedocs.io/en/latest/contributing.html#pull-requests)
- [x] Add entry to the [release notes](https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by cburgdorf** [comment]
The [remaining error](https://circleci.com/gh/ethereum/trinity/230041?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) is interesting, the `receipt_root` (and hence also the `state_root`) mismatch but the same test runs fine in `Py-EVM`. :thinking:

**Comment by cburgdorf** [comment]
Here's a trace of Py-EVM successfully running that fixture via:
 `pytest tests/json-fixtures/test_blockchain.py -k 'chainId_d0g0v0_Istanbul' --log-cli-level 0`

```
-------------------------------------------------------------------------------------------------- live log call --------------------------------------------------------------------------------------------------
  DEBUG2  05-27 15:28:14            logging.py  Persisting new state root: 0xaff9f63320a482f8c4e4f15f659e6a7ac382138fbbb6919243b0cba4c5988a5a
  DEBUG2  05-27 15:28:14            logging.py  TRANSACTION: sender: 0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b | to: 0x1000000000000000000000000000000000000000 | value: 0 | gas: 100000000 | gas-price: 1 | s: 48306522585199634021120931500637254620582923582825676399052146585129481897754 | r: 72166270929973037356569147027147722002648387518809749561219742230937616907299 | v: 27 | data-hash: 0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470
  DEBUG2  05-27 15:28:14            logging.py  COMPUTATION STARTING: gas: 99979000 | from: 0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b | to: 0x1000000000000000000000000000000000000000 | value: 0 | depth 0 | static: n
  DEBUG2  05-27 15:28:14            logging.py  OPCODE: 0x46 (CHAINID) | pc: 0
  DEBUG2  05-27 15:28:14            logging.py  GAS CONSUMPTION: 99979000 - 2 -> 99978998 (CHAINID)
  DEBUG2  05-27 15:28:14            logging.py  OPCODE: 0x60 (PUSH1) | pc: 1
  DEBUG2  05-27 15:28:14            logging.py  GAS CONSUMPTION: 99978998 - 3 -> 99978995 (PUSH1)
  DEBUG2  05-27 15:28:14            logging.py  OPCODE: 0x55 (SSTORE) | pc: 3
  DEBUG2  05-27 15:28:14            logging.py  GAS CONSUMPTION: 99978995 - 20000 -> 99958995 (SSTORE: 0x1000000000000000000000000000000000000000[1] -> 1 (current: 0 / original: 0))
  DEBUG2  05-27 15:28:14            logging.py  OPCODE: 0x0 (STOP) | pc: 4
  DEBUG2  05-27 15:28:14            logging.py  COMPUTATION SUCCESS: from: 0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b | to: 0x1000000000000000000000000000000000000000 | value: 0 | depth: 0 | static: n | gas-used: 20005 | gas-remaining: 99958995
  DEBUG2  05-27 15:28:14            logging.py  TRANSACTION REFUND: 99958995 -> 0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b
  DEBUG2  05-27 15:28:14            logging.py  TRANSACTION FEE: 41005 -> 0x2adc25665018aa1fe0e6bc666dac8fc2697ff9ba
   DEBUG  05-27 15:28:14               base.py  BLOCK REWARD: 2000000000000000000 -> b'*\xdc%fP\x18\xaa\x1f\xe0\xe6\xbcfm\xac\x8f\xc2i\x7f\xf9\xba'
  DEBUG2  05-27 15:28:14            logging.py  Updating account 0x1000000000000000000000000000000000000000 to storage root 0xf38f9f63c760d088d7dd04f743619b6291f63beebd8bdf530628f90e9cfa52d7
  DEBUG2  05-27 15:28:14            logging.py  persist storage root to data store
  DEBUG2  05-27 15:28:14            logging.py  Persisting new state root: 0x9c736f76000cc7d20fdc892c260c08a545c6d6acae8edcf9cb214c7049d5aa80
   DEBUG  05-27 15:28:14               base.py  Block #1-0x590a..8317 reads 4 unique node hashes, 3 addresses, 1 bytecodes, and 1 storage slots
```

And here's a trace of Trinity running that same fixture via:
 `pytest tests/json-fixtures-over-rpc/test_rpc_fixtures.py --fork Istanbul -k 'chainId_d0g0v0_Istanbul' -vv --log-cli-level 0`

```
-------------------------------------------------------------------------------------------------- live log call --------------------------------------------------------------------------------------------------
  DEBUG2  05-27 15:34:45            logging.py  Persisting new state root: 0xaff9f63320a482f8c4e4f15f659e6a7ac382138fbbb6919243b0cba4c5988a5a
  DEBUG2  05-27 15:34:45            logging.py  TRANSACTION: sender: 0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b | to: 0x1000000000000000000000000000000000000000 | value: 0 | gas: 100000000 | gas-price: 1 | s: 48306522585199634021120931500637254620582923582825676399052146585129481897754 | r: 72166270929973037356569147027147722002648387518809749561219742230937616907299 | v: 27 | data-hash: 0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470
  DEBUG2  05-27 15:34:45            logging.py  COMPUTATION STARTING: gas: 99979000 | from: 0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b | to: 0x1000000000000000000000000000000000000000 | value: 0 | depth 0 | static: n
  DEBUG2  05-27 15:34:45            logging.py  OPCODE: 0x46 (CHAINID) | pc: 0
  DEBUG2  05-27 15:34:45            logging.py  GAS CONSUMPTION: 99979000 - 2 -> 99978998 (CHAINID)
  DEBUG2  05-27 15:34:45            logging.py  OPCODE: 0x60 (PUSH1) | pc: 1
  DEBUG2  05-27 15:34:45            logging.py  GAS CONSUMPTION: 99978998 - 3 -> 99978995 (PUSH1)
  DEBUG2  05-27 15:34:45            logging.py  OPCODE: 0x55 (SSTORE) | pc: 3
  DEBUG2  05-27 15:34:45            logging.py  GAS CONSUMPTION: 99978995 - 800 -> 99978195 (SSTORE: 0x1000000000000000000000000000000000000000[1] -> 0 (current: 0 / original: 0))
  DEBUG2  05-27 15:34:45            logging.py  OPCODE: 0x0 (STOP) | pc: 4
  DEBUG2  05-27 15:34:45            logging.py  COMPUTATION SUCCESS: from: 0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b | to: 0x1000000000000000000000000000000000000000 | value: 0 | depth: 0 | static: n | gas-used: 805 | gas-remaining: 99978195
  DEBUG2  05-27 15:34:45            logging.py  TRANSACTION REFUND: 99978195 -> 0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b
  DEBUG2  05-27 15:34:45            logging.py  TRANSACTION FEE: 21805 -> 0x2adc25665018aa1fe0e6bc666dac8fc2697ff9ba
   DEBUG  05-27 15:34:45               base.py  BLOCK REWARD: 2000000000000000000 -> b'*\xdc%fP\x18\xaa\x1f\xe0\xe6\xbcfm\xac\x8f\xc2i\x7f\xf9\xba'
  DEBUG2  05-27 15:34:45            logging.py  Persisting new state root: 0xc74c990c2db5818d81d9c5dd7f8d78ad6ecd54bb5ffdfa21fa4c331099163195
   DEBUG  05-27 15:34:45               base.py  Block #1-0xcad0..a7a2 reads 4 unique node hashes, 3 addresses, 1 bytecodes, and 1 storage slots
```

Clearly we are charging a different `SSTORE` and I believe @carver was deep into similar issues some weeks ago. Any idea what might be going on here? Also summoning @veox as this might be your kind of fun :sweat_smile:

File: tests/json-fixtures-over-rpc/test_rpc_fixtures.py [hunk 1]
```diff
@@ -135,6 +136,7 @@
     # * https://github.com/ethereum/py-evm/pull/1181#issuecomment-446330609
     ('GeneralStateTests/stCreate2/RevertInCreateInInitCreate2_d0g0v0.json', 'RevertInCreateInInitCreate2_d0g0v0_Constantinople'),  # noqa: E501
     ('GeneralStateTests/stCreate2/RevertInCreateInInitCreate2_d0g0v0.json', 'RevertInCreateInInitCreate2_d0g0v0_ConstantinopleFix'),  # noqa: E501
+    ('GeneralStateTests/stCreate2/RevertInCreateInInitCreate2.json', 'RevertInCreateInInitCreate2_d0g0v0_Istanbul'),  # noqa: E501
```
Comment (carver) [inline tests/json-fixtures-over-rpc/test_rpc_fixtures.py:139]: Interesting, the name pattern changed? By pattern matching, I would have expected:
```diff
-    ('GeneralStateTests/stCreate2/RevertInCreateInInitCreate2.json', 'RevertInCreateInInitCreate2_d0g0v0_Istanbul'),  # noqa: E501
+    ('GeneralStateTests/stCreate2/RevertInCreateInInitCreate2_d0g0v0.json', 'RevertInCreateInInitCreate2_d0g0v0_Istanbul'),  # noqa: E501
```

File: trinity/rpc/format.py [hunk 2]
```diff
@@ -96,7 +95,11 @@ def to_receipt_response(receipt: ReceiptAPI,
         ],
         "logsBloom": format_bloom(receipt.bloom),
         "root": encode_hex(receipt.state_root),
-        "to": encode_hex(transaction.to),
+        "to": apply_formatter_if(
+            is_address,
+            to_checksum_address,
+            encode_hex(transaction.to)
+        ),
```
Comment (carver) [inline trinity/rpc/format.py:102]: Having the encode in the argument tripped me for a second. I took a stab at an alternative.

```suggestion
        "to": compose(
            apply_formatter_if(
                is_address,
                to_checksum_address
            ),
            encode_hex,
        )(transaction.to),
```

Not really sure it's any better.

**Comment by carver** [comment]
lol whoops, went straight to the code, missed the comments, and didn't realize there was still a broken test.

**Comment by carver** [comment]
> but the same test runs fine in `Py-EVM`.

Was that running against `py-evm==0.3.0a15`?

**Comment by carver** [comment]
Are you *really* certain those are the same tests running? The py-evm one seems to have the first `SSTORE` set the value to 1, but the trinity one is using `SSTORE` to set the value to 0.

py-evm:
```
  DEBUG2  05-27 15:28:14            logging.py  GAS CONSUMPTION: 99978995 - 20000 -> 99958995 (SSTORE: 0x1000000000000000000000000000000000000000[1] -> 1 (current: 0 / original: 0))
```

trinity:
```
  DEBUG2  05-27 15:34:45            logging.py  GAS CONSUMPTION: 99978995 - 800 -> 99978195 (SSTORE: 0x1000000000000000000000000000000000000000[1] -> 0 (current: 0 / original: 0))
```

**Comment by cburgdorf** [comment]
>Are you really certain those are the same tests running?

Yeah, they should be. Also from debugging I can tell that both are attempting to persist the very same identical block but when the test is run through Trinity, the `receipt_root` changes after the import and we run into the validation error.

```
  DEBUG2  05-27 19:41:02            logging.py  Persisting new state root: 0xc74c990c2db5818d81d9c5dd7f8d78ad6ecd54bb5ffdfa21fa4c331099163195
   DEBUG  05-27 19:41:02               base.py  Block #1-0xcad0..a7a2 reads 4 unique node hashes, 3 addresses, 1 bytecodes, and 1 storage slots
 WARNING  05-27 19:45:25               base.py  Proposed Block #1-0x590a..8317 doesn't follow EVM rules, rejecting...
   DEBUG  05-27 19:45:25               main.py  Validation error while executing RPC method
Traceback (most recent call last):
  File "/home/cburgdorf/Documents/hacking/ef/trinity/trinity/rpc/main.py", line 120, in _get_result
    result = await execute_with_retries(
  File "/home/cburgdorf/Documents/hacking/ef/trinity/trinity/rpc/retry.py", line 108, in execute_with_retries
    return await func(*params)
  File "/home/cburgdorf/Documents/hacking/ef/trinity/trinity/rpc/format.py", line 213, in formatted_func
    return await func(self, *formatted)
  File "/home/cburgdorf/Documents/hacking/ef/trinity/trinity/rpc/modules/evm.py", line 57, in applyBlockFixture
    _, _, rlp_encoded = apply_fixture_block_to_chain(block_info, self.chain)
  File "/home/cburgdorf/Documents/hacking/ef/py-evm/eth/tools/fixtures/helpers.py", line 222, in apply_fixture_block_to_chain
    import_result = chain.import_block(block, perform_validation=perform_validation)
  File "/home/cburgdorf/Documents/hacking/ef/py-evm/eth/chains/base.py", line 475, in import_block
    if perform_validation:
  File "cytoolz/functoolz.pyx", line 250, in cytoolz.functoolz.curry.__call__
  File "/home/cburgdorf/Documents/hacking/ef/py-evm/eth/_utils/rlp.py", line 89, in validate_rlp_equal
    raise ValidationError(error_message)
eth_utils.exceptions.ValidationError: Mismatch between block and imported block on 3 fields:
 - header.state_root  :
    (actual)  : b'\xc7L\x99\x0c-\xb5\x81\x8d\x81\xd9\xc5\xdd\x7f\x8dx\xadn\xcdT\xbb_\xfd\xfa!\xfaL3\x10\x99\x161\x95'
    (expected): b'\x9csov\x00\x0c\xc7\xd2\x0f\xdc\x89,&\x0c\x08\xa5E\xc6\xd6\xac\xae\x8e\xdc\xf9\xcb!LpI\xd5\xaa\x80'
 - header.receipt_root:
    (actual)  : b'\xc8\x00\xa1\x07\xfcH\x0b\xf0\xc7i\x08\x1d\xf7\x96\xe0w\x8c\x00\x06\xf2\xc7m>B\x021\xf4\xe4\xa6\xc0\xb8\xdb'
    (expected): b'\xfa\x9e\x94,{\xab\x10\x17\xc2\x9a\xb8\xb7\xf9HN1\x1f:+\xa6\x80\xc2\xec\x8a\xbb\xae\xa26\\\xec\xc9>'
 - header.gas_used    :
    (actual)  : 21805
    (expected): 41005
```

**Comment by cburgdorf** [comment]
Anyway, I'm at the end of my work day. I'll continue debugging this tomorrow.

**Comment by cburgdorf** [comment]
Ok, I've found the issue. The test is pushing the `chain_id` on the stack and if the test runs through trinity, it uses a different `chain_id` than it should be using. Guess that shouldn't be too hard to fix.

File: trinity/rpc/format.py (hunk 2 above)
Comment (cburgdorf) [reply trinity/rpc/format.py:102]: It was already used that way in `transaction_to_dict` and I personally find it slightly more readable but I'd be fine either way.

**Comment by cburgdorf** [comment]
Ok, so the problem was that our `MainnetFullChain` (the one that is the `MainnetChain` + Trinitys async mixins) did not have the `chain_id` set which tripped up this test as it is testing the `CHAINID` opcode which ended up being `0` (default `int`) when it should be `1`.

File: tests/json-fixtures-over-rpc/test_rpc_fixtures.py (hunk 1 above)
Comment (cburgdorf) [reply tests/json-fixtures-over-rpc/test_rpc_fixtures.py:140]: I double checked it with the Py-EVM test and it seems to be correct though.

**Pull request merged**